Homework 1

zavijavasoft opened the pull request:
Пулл-реквест на первую домашку

Commit:
Добавлены интерфейсы и абстрактные классы для бизнес-логики

Добавлены
1.Сущность транзакции (приход/расход)
2.Интерфейс для обращения к источнику данных транзакций
3.Интерфейс для доступа к списку валют
4.Абстрактный класс, реализующий интерфейс для мониторинга курсов валют

Commit:
Добавляем непосредственно сам финансовый трекер для расчета баланса

Трекер реализуем в виде объекта-одиночки, который зависит только от абстрактных интерфейсов и классов данных предыдущего коммита

Commit:
Добавляем объекты-заглушки для источников данных

На данном этапе мы работаем с "хардкодными" данными, которые и предоставляются этими заглушками

Commit:
Добавляем тесты для объектов-заглушек

Интересуют тесты не столько самих заглушек, сколько тесты нетривиальной логики базового класса AbstractCurrencyMonitor.
Плюс надо было удостовериться, что я правильно понимаю работу наследования и переопределения в Kotlin

Commit:
Добавляем тесты для подсчета баланса

Тесты проверяются на хардкодных данных

Commit:
Модель переделана для упрощения

Commit:
Папка core переименована в model

Commit:
Добавлены ресурсы

Commit:
Доведен до ума сам финансовый трекер, добавлен доп метод

 Скорректированы тесты

Commit:
Добавлены модели и компоненты для DI

Заодно переименовал все интерфейсы в одну нотацию

Commit:
Добавлены элементы презентера-представления для первой вкладки баланса

Commit:
Добавлены элементы представления для  вкладки About

Commit:
Добавлены элементы  презентера-представления для  вкладки настроек

в этот же коммит добавлю заглушки для списка транзакций

Commit:
Настройка Dagger

Commit:
Добавлен презентер главного активити

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/YaFinaActivity.kt:
```diff
@@ -0,0 +1,115 @@
+    lateinit var mainPresenter: MainPresenter
+
+
+    lateinit var _fragmentManager: FragmentManager
```
зачем заводить отдельную переменную?

zavijavasoft replied:
fixed

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/YaFinaActivity.kt:
```diff
@@ -0,0 +1,115 @@
+    lateinit var _fragmentManager: FragmentManager
+    var tabState: MainTabs = MainTabs.BALANCE
+
+    val balanceFragment = BalanceFragment.getInstance()
```
При старте этой активити ты получается сразу создаешь 3 фрагмента. А использоваться может потом только один.

При повороте у тебя создается еще 3 фрагмента. При этом система сама восстановит текущий показанный фрагмент. Значит у тебя будет 4 фрагмента.

Лучше не держать ссылок на эти фрагменты и в нужном месте создавать инстанс когда потребуется.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/YaFinaActivity.kt:
```diff
@@ -0,0 +1,115 @@
+
+            if (fragment == null) {
+
+                val fragmentTransaction = _fragmentManager.beginTransaction()
```
обычно если не нужно делать каких-то if(...), то методы транзакции можно чейнить без заведения переменной.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/core/BalancePresenter.kt:
```diff
@@ -0,0 +1,35 @@
+
+
+    init {
+        YaFinaApplication.component.inject(this)
```
Я не работал с Moxy, может там так и принятно, то из-за этого теряется следующее удобство:

Получается твой презентер зависит от даггера и что бы его тестить нужно инициализировать даггер и подсовывать зависимости через него.

Если сделать передачу зависимостей через конструктор презентера, то unit tests можно писать для этого класса вообще не используя даггер - передал моки в конструктор и все.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/core/IBalancePresenter.kt:
```diff
@@ -0,0 +1,5 @@
+package com.zavijavasoft.yafina.core
+
+interface IBalancePresenter {
```
Хорошо - когда именования в проекте выполнены в одном стиле и подчиняются каким-то правилам. Но все же в java обычно интерфейсы именуют без префикса I (в отличии от шарпа).

Но это скорее рекомендация, чем жесткое правило.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/core/SettingsPresenter.kt:
```diff
@@ -0,0 +1,33 @@
+
+    override fun addCurrency(currency: String) {
+        val list = storage.getCurrencyList()
+        if (currency !in list)
```
Опять же рекомендация: писать {} всегда. Это может избавить от нежелательных проблем в дальнейшем, когда ты добавишь строчку в if, а скобки поставить забудешь.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/FinanceTrackerImpl.kt:
```diff
@@ -0,0 +1,74 @@
+
+class FinanceTrackerImpl(var storage: ITransactionStorage) : IFinanceTracker {
+
+    private var _transactions: List<TransactionInfo> = listOf()
```
По идее тут можно обойтись менем transactions.
Темболее именование начинающееся с _ сейчас не очень приняты.

zavijavasoft replied:
_transactions -- это реализация оверрайденного transactions из интерфейса.

HotIceCream replied:
А так получится? 
```
    override var transactions: List<TransactionInfo> = listOf()
        private set
```

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/FinanceTrackerImpl.kt:
```diff
@@ -0,0 +1,74 @@
+import com.zavijavasoft.yafina.utils.roundSum
+
+
+class FinanceTrackerImpl(var storage: ITransactionStorage) : IFinanceTracker {
```
Вот лучше именовать так: FinanceTrackerImplt : FinanceTracker

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/FinanceTrackerImpl.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+
+    override fun calculateBalance(currency: String, trans: List<TransactionInfo>): Float {
```
Есть еще хорошее правило именования: никогда не использовать сокращения. Код, который ты пишешь будет читать другой человек и он не всегда поймет что за трансы тут используются.

Даже ты сам через некоторое время можешь забыть как называл переменную: trans, transacts, transactions. 
Поэтому следуюешь правилу - без сокращений - и всегда знаешь как называется переменная.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/ui/BalanceFragment.kt:
```diff
@@ -0,0 +1,69 @@
+                              savedInstanceState: Bundle?): View? {
+
+        val view = inflater.inflate(R.layout.fragment_balance, container, false)
+        ButterKnife.bind(this, view)
```
у ButterKnife есть фича - вот тут возвращается Unbinder.
Т.к. у одного фрагмента View может создаваться несколько раз, то стоит сохранять Unbinder и в onDestroyView вызывать его.

Правда я не знаю как он будет работать с котлином =\

Ну либо знать, что твой фрагмент не будет использоваться с таким lifecycle.

HotIceCream commented on app/src/test/java/com/zavijavasoft/yafina/StubTest.kt:
```diff
@@ -0,0 +1,32 @@
+
+class StubTest {
+
+    var i = 0
```
?

HotIceCream commented on app/src/test/java/com/zavijavasoft/yafina/IFinanceTrackerImplTest.kt:
```diff
@@ -0,0 +1,59 @@
+import org.junit.Test
+import java.math.BigDecimal
+
+class IFinanceTrackerImplTest {
```
А тут именование смешано

HotIceCream commented on app/src/main/res/layout/fragment_balance.xml:
```diff
@@ -0,0 +1,71 @@
+        android:textAlignment="center"
+        android:textColor="@color/colorYaYellow"
+        android:textSize="25sp"
+        app:layout_constraintBottom_toTopOf="@id/balance_rur"
```
Эти атрибуты тут еще нужны?

HotIceCream reviewed (approved):
В целом все ок.

Основные замечания у меня по стилистике кода

Commit:
Исправление замечаний CR по YaFinaActivity

Commit:
Исправление замечаний CR по тестам

Commit:
Исправление замечаний CR по I-нотации

Также устранено замечание по трансу:)

Commit:
Исправление косметического замечания CR

Commit:
Исправление самого серьезного косяка, замеченного при CR

Путем жесткого перепахивания кода зависимость презентеров от даггера была убрана.
Теперь ижектирование ведется во фрагментах (пока в одном)
Заодно наведен порядок в модулях даггера и тестовые классы переименованы под
однородную нотацию

Commit:
Исправление последнего замечания CR

Обработка unbinder'ов ButterKnife
